azure/src/main/java/org/apache/iceberg/azure/adlsv2/ADLSInputStream.java

```diff
@@ -0,0 +1,203 @@
+import com.azure.storage.file.datalake.models.FileRange;
+import com.azure.storage.file.datalake.options.DataLakeFileInputStreamOptions;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import org.apache.iceberg.azure.AzureProperties;
+import org.apache.iceberg.io.FileIOMetricsContext;
+import org.apache.iceberg.io.IOUtil;
+import org.apache.iceberg.io.RangeReadable;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.metrics.Counter;
+import org.apache.iceberg.metrics.MetricsContext;
+import org.apache.iceberg.metrics.MetricsContext.Unit;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.io.ByteStreams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ADLSInputStream extends SeekableInputStream implements RangeReadable {
+  private static final Logger LOG = LoggerFactory.getLogger(ADLSInputStream.class);
+
+  // max skip size, if skip is larger, then the stream will be reopened at new position
+  private static final int SKIP_SIZE = 1024 * 1024;
+
+  private final StackTraceElement[] createStack;
+  private final DataLakeFileClient fileClient;
+  private Long fileSize;
+  private final AzureProperties azureProperties;
+
+  private InputStream stream;
+  private long pos;
+  private long next;
+  private boolean closed;
+
+  private final Counter readBytes;
+  private final Counter readOperations;
+
+  ADLSInputStream(
+      DataLakeFileClient fileClient,
+      Long fileSize,
+      AzureProperties azureProperties,
+      MetricsContext metrics) {
+    this.fileClient = fileClient;
+    this.fileSize = fileSize;
+    this.azureProperties = azureProperties;
+
+    this.readBytes = metrics.counter(FileIOMetricsContext.READ_BYTES, Unit.BYTES);
+    this.readOperations = metrics.counter(FileIOMetricsContext.READ_OPERATIONS);
+
+    this.createStack = Thread.currentThread().getStackTrace();
+
+    openStream();
+  }
+
+  private void openStream() {
+    DataLakeFileOpenInputStreamResult result =
+        fileClient.openInputStream(getInputOptions(new FileRange(pos)));
+    this.fileSize = result.getProperties().getFileSize();
+    this.stream = result.getInputStream();
+  }
+
+  private DataLakeFileInputStreamOptions getInputOptions(FileRange range) {
+    DataLakeFileInputStreamOptions options = new DataLakeFileInputStreamOptions();
+    azureProperties.adlsReadBlockSize().ifPresent(options::setBlockSize);
+    options.setRange(range);
+    return options;
+  }
+
+  @Override
+  public long getPos() {
+    return next;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "Cannot seek: already closed");
+    Preconditions.checkArgument(newPos >= 0, "Cannot seek: position %s is negative", newPos);
+
+    // this allows a seek beyond the end of the stream but the next read will fail
```

Comment:
Why allow seek beyond the end of the stream?

Reply:
This was done to keep the behavior consistent with [S3InputStream](https://github.com/apache/iceberg/blob/3ab00171b48bb793a3b71845eb12d5077ba892f1/aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java#L87)